File: fetch-silver.js
// fetch-silver.js
// Fetches silver price (per tola) from fenegosida.org
// Runs via GitHub Actions twice daily — 10:30 AM + 11:10 AM NPT

const https = require('https');

const PROJECT_ID = process.env.FIREBASE_PROJECT_ID;
const API_KEY    = process.env.FIREBASE_API_KEY;

if (!PROJECT_ID || !API_KEY) {
  console.error('❌ Missing FIREBASE_PROJECT_ID or FIREBASE_API_KEY env vars.');
  process.exit(1);
}

// ── Fetch URL ──
function get(url) {
  return new Promise((resolve, reject) => {
    const req = https.get(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Accept':     'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.5',
        'Cache-Control':   'no-cache',
      }
    }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return get(res.headers.location).then(resolve).catch(reject);
      }
      if (res.statusCode !== 200) {
        return reject(new Error('HTTP ' + res.statusCode));
      }
      let data = '';
      res.setEncoding('utf8');
      res.on('data', chunk => data += chunk);
      res.on('end', () => resolve(data));
    });
    req.on('error', reject);
    req.setTimeout(20000, () => { req.destroy(); reject(new Error('Timeout')); });
  });
}

// ── Multi-strategy fetch (bypass IP blocking) ──
async function fetchFenegosida() {
  const TARGET = 'https://fenegosida.org/';
  const strategies = [
    { name: 'Direct',      url: TARGET },
    { name: 'allorigins',  url: 'https://api.allorigins.win/raw?url=' + encodeURIComponent(TARGET) },
    { name: 'corsproxy',   url: 'https://corsproxy.io/?' + encodeURIComponent(TARGET) },
    { name: 'codetabs',    url: 'https://api.codetabs.com/v1/proxy?quest=' + encodeURIComponent(TARGET) },
    { name: 'thingproxy',  url: 'https://thingproxy.freeboard.io/fetch/' + TARGET },
  ];
  for (const s of strategies) {
    try {
      console.log('🌐 Trying:', s.name);
      const html = await get(s.url);
      if (html && html.includes('SILVER')) {
        console.log('✅', s.name, 'succeeded');
        return html;
      }
    } catch (e) {
      console.log('⚠️ ', s.name, 'failed:', e.message);
    }
  }
  throw new Error('All fetch strategies failed.');
}

// ── Strip HTML tags/scripts/styles down to plain text ──
// Key fix: the old parser depended on an exact "* tola" marker string and
// raw <b> tag positions, both of which are extremely sensitive to the
// site's markup. Parsing plain, tag-free text instead makes this resistant
// to markup changes as long as the visible wording stays roughly the same.
function stripHtml(html) {
  return html
    .replace(/<script[\s\S]*?<\/script>/gi, ' ')
    .replace(/<style[\s\S]*?<\/style>/gi, ' ')
    .replace(/<!--[\s\S]*?-->/g, ' ')
    .replace(/<[^>]+>/g, ' ')
    .replace(/&nbsp;/gi, ' ')
    .replace(/&#2352;/g, '\u0930')   // रु rendered as numeric HTML entities, just in case
    .replace(/&#2369;/g, '\u0941')
    .replace(/\s+/g, ' ')
    .trim();
}

// ── Parse silver price (per tola) from fenegosida.org ──
//
// Page lists FINE GOLD, TEJABI GOLD, SILVER, each with a "per 10 grm" row
// and a "per 1 tola" row. We want SILVER's "per 1 tola" figure — anchor
// directly on the phrase "SILVER ... per 1 tola" (not "per 10 grm"), then
// grab the रु/Nrs-prefixed price shortly after it.
function parseSilverPrice(html) {
  const text = stripHtml(html);

  // Primary strategy: anchor on "SILVER" immediately followed by "per 1 tola"
  const m = text.match(/SILVER\s*per\s*1\s*tola[\s\S]{0,200}?(?:[\u0930\u0941]+|Nrs)\s*\*{0,2}(\d{3,6})(?:\.\d+)?\*{0,2}/i);
  if (m) {
    const price = parseInt(m[1], 10);
    if (price > 500 && price < 50000) {
      console.log('✅ Anchored SILVER per-1-tola price found:', price);
      return price;
    }
  }

  // Fallback 1: same anchor but with a wider window, in case more text sits between
  const wide = text.match(/SILVER\s*per\s*1\s*tola[\s\S]{0,600}?(?:[\u0930\u0941]+|Nrs)\s*\*{0,2}(\d{3,6})(?:\.\d+)?\*{0,2}/i);
  if (wide) {
    const price = parseInt(wide[1], 10);
    if (price > 500 && price < 50000) {
      console.log('⚠️  Wide-window SILVER per-1-tola price found:', price);
      return price;
    }
  }

  // Fallback 2: "per 1 tola" occurs 3x on the page (Fine Gold, Tejabi, Silver, in that
  // order) — Silver is always the third occurrence, so grab the price right after it.
  const tolaMatches = [...text.matchAll(/per\s*1\s*tola/gi)];
  if (tolaMatches.length >= 3) {
    const idx = tolaMatches[2].index + tolaMatches[2][0].length;
    const window = text.slice(idx, idx + 200);
    const wm = window.match(/(?:[\u0930\u0941]+|Nrs)\s*\*{0,2}(\d{3,6})(?:\.\d+)?\*{0,2}/i);
    if (wm) {
      const price = parseInt(wm[1], 10);
      if (price > 500 && price < 50000) {
        console.log('⚠️  Third per-1-tola occurrence price found:', price);
        return price;
      }
    }
  }

  // Fallback 3: any 3-6 digit number in silver range after the last "tola" keyword
  const tolaBlock = text.match(/tola[\s\S]{0,1000}$/i);
  if (tolaBlock) {
    const nums = [...tolaBlock[0].matchAll(/\b(\d{3,5})(?:\.\d+)?\b/g)].map(m => parseInt(m[1]));
    const valid = nums.filter(n => n > 500 && n < 50000);
    if (valid.length) {
      console.log('⚠️  Silver tola block fallback:', valid[valid.length - 1]);
      return valid[valid.length - 1];
    }
  }

  return null;
}

// ── Write to global_data/silver_info ──
function writeToFirestore(price) {
  return new Promise((resolve, reject) => {
    const body = JSON.stringify({
      fields: {
        rate:        { integerValue: String(price) },
        updatedBy:   { stringValue: 'auto@github-actions' },
        lastUpdated: { stringValue: new Date().toISOString() },
        source:      { stringValue: 'fenegosida.org' },
        note:        { stringValue: 'Silver per tola' }
      }
    });
    const urlObj = new URL(
      `https://firestore.googleapis.com/v1/projects/${PROJECT_ID}/databases/(default)/documents/global_data/silver_info?key=${API_KEY}`
    );
    const req = https.request({
      hostname: urlObj.hostname,
      path:     urlObj.pathname + urlObj.search,
      method:   'PATCH',
      headers: {
        'Content-Type':   'application/json',
        'Content-Length': Buffer.byteLength(body)
      }
    }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        if (res.statusCode === 200) resolve(JSON.parse(data));
        else reject(new Error('Firestore error ' + res.statusCode + ': ' + data));
      });
    });
    req.on('error', reject);
    req.write(body);
    req.end();
  });
}

// ── Log to silver_history collection ──
function logToHistory(price) {
  return new Promise((resolve) => {
    const body = JSON.stringify({
      fields: {
        rate:      { integerValue: String(price) },
        updatedBy: { stringValue: 'auto@github-actions' },
        timestamp: { stringValue: new Date().toISOString() },
        source:    { stringValue: 'fenegosida.org' },
        note:      { stringValue: 'Silver per tola' }
      }
    });
    const urlObj = new URL(
      `https://firestore.googleapis.com/v1/projects/${PROJECT_ID}/databases/(default)/documents/silver_history?key=${API_KEY}`
    );
    const req = https.request({
      hostname: urlObj.hostname,
      path:     urlObj.pathname + urlObj.search,
      method:   'POST',
      headers: {
        'Content-Type':   'application/json',
        'Content-Length': Buffer.byteLength(body)
      }
    }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => resolve(data));
    });
    req.on('error', e => resolve('history log skipped: ' + e.message));
    req.write(body);
    req.end();
  });
}

// ── Main ──
async function main() {
  console.log('🥈 Silver price auto-fetch starting...');
  console.log('📅 Time (UTC):', new Date().toISOString());
  console.log('🌐 Source: fenegosida.org');

  let html = null;
  try {
    html = await fetchFenegosida();
    console.log('📄 Page size:', Math.round(html.length / 1024) + 'KB');
  } catch (e) {
    console.error('❌ Fetch failed:', e.message);
    process.exit(1);
  }

  const price = parseSilverPrice(html);
  if (!price) {
    console.error('❌ Could not parse silver price from fenegosida.org');
    process.exit(1);
  }

  console.log('🥈 Silver price (per tola): NPR', price.toLocaleString());

  try {
    await writeToFirestore(price);
    console.log('🔥 Silver written to Firestore → global_data/silver_info');
  } catch (e) {
    console.error('❌ Firestore write failed:', e.message);
    process.exit(1);
  }

  try {
    await logToHistory(price);
    console.log('📋 Silver logged to silver_history collection');
  } catch (e) {
    console.log('⚠️  History log skipped:', e.message);
  }

  console.log('');
  console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  console.log('✅ Done! silver_info.rate =', price);
  console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
}

main();

File: fetch_silver.py
#!/usr/bin/env python3
"""
fetch_silver.py
Fetches Silver price per tola for Nepal and writes it to Firebase Firestore.
Runs via GitHub Actions twice daily — 10:30 AM + 11:10 AM NPT.

Same source-chain design as fetch_gold.py — see that file's docstring for why.
This version also carries over the proxy-fallback layer from the original
fetch-silver.js (fenegosida.org direct fetches were sometimes IP-blocked,
separately from the JS-rendering problem), so within the fenegosida.org
source itself there are several fetch attempts before giving up on it.
"""

import os
import re
import sys
import json
import time
import urllib.request
import urllib.error
import urllib.parse
from datetime import datetime, timezone

# ── Firebase config (injected from GitHub Secrets) ──
PROJECT_ID = os.environ.get("FIREBASE_PROJECT_ID")
API_KEY = os.environ.get("FIREBASE_API_KEY")

if not PROJECT_ID or not API_KEY:
    print("❌ Missing FIREBASE_PROJECT_ID or FIREBASE_API_KEY env vars.")
    sys.exit(1)

# Sane bounds for a silver price per tola, in NPR.
MIN_PRICE = 500
MAX_PRICE = 50_000

HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/124.0 Safari/537.36 SilverBot/2.0"
    ),
    "Accept": "text/html,application/xhtml+xml,application/json",
    "Accept-Language": "en-US,en;q=0.5",
    "Cache-Control": "no-cache",
}


# ───────────────────────── HTTP helpers ─────────────────────────

def fetch_url(url, timeout=15, retries=2, backoff=2):
    """GET a URL with retries + exponential backoff. Returns response text or None."""
    last_err = None
    for attempt in range(1, retries + 1):
        try:
            req = urllib.request.Request(url, headers=HEADERS)
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                if resp.status != 200:
                    raise urllib.error.HTTPError(url, resp.status, "bad status", None, None)
                return resp.read().decode("utf-8", errors="replace")
        except Exception as e:  # noqa: BLE001 - want to catch+retry everything network related
            last_err = e
            print(f"      ⚠️  attempt {attempt}/{retries} failed: {e}")
            if attempt < retries:
                time.sleep(backoff ** attempt)
    print(f"      ❌ all {retries} attempts failed: {last_err}")
    return None


def strip_html(html):
    """Collapse HTML down to plain, whitespace-normalized text."""
    text = re.sub(r"<script[\s\S]*?</script>", " ", html, flags=re.I)
    text = re.sub(r"<style[\s\S]*?</style>", " ", text, flags=re.I)
    text = re.sub(r"<!--[\s\S]*?-->", " ", text)
    text = re.sub(r"<[^>]+>", " ", text)
    text = text.replace("&nbsp;", " ")
    text = text.replace("&#2352;", "\u0930")  # र
    text = text.replace("&#2369;", "\u0941")  # ु
    text = re.sub(r"\s+", " ", text).strip()
    return text


def in_range(n):
    return MIN_PRICE < n < MAX_PRICE


# ───────────────────────── Source 1: fenegosida.org (+ proxy fallbacks) ─────────────────────────
# fenegosida.org is now a client-rendered JS SPA, so a direct fetch usually
# returns an empty shell with no price text — the proxies below don't fix
# that (they can't execute JS either), but they're kept because the original
# script needed them for straight IP-blocking issues independent of
# rendering, and this way that layer isn't lost if the site's rendering
# situation ever changes.

def build_proxy_targets():
    target = "https://fenegosida.org/"
    encoded = urllib.parse.quote(target, safe="")
    return [
        ("Direct", target),
        ("allorigins", f"https://api.allorigins.win/raw?url={encoded}"),
        ("corsproxy", f"https://corsproxy.io/?{encoded}"),
        ("codetabs", f"https://api.codetabs.com/v1/proxy?quest={encoded}"),
        ("thingproxy", f"https://thingproxy.freeboard.io/fetch/{target}"),
    ]


def parse_embedded_json(html):
    patterns = [
        r'<script[^>]*id=["\']__NEXT_DATA__["\'][^>]*>([\s\S]*?)</script>',
        r'window\.__INITIAL_STATE__\s*=\s*(\{[\s\S]*?\});',
        r'window\.__NUXT__\s*=\s*(\{[\s\S]*?\});',
        r'<script[^>]*type=["\']application/json["\'][^>]*>([\s\S]*?)</script>',
    ]
    for pat in patterns:
        for m in re.finditer(pat, html, flags=re.I):
            blob = m.group(1)
            try:
                data = json.loads(blob)
            except (json.JSONDecodeError, TypeError):
                continue
            price = search_json_for_silver_price(data)
            if price:
                return price
    return None


def search_json_for_silver_price(obj, depth=0):
    if depth > 6:
        return None
    if isinstance(obj, dict):
        for key, val in obj.items():
            key_l = str(key).lower()
            if isinstance(val, (int, float)) and in_range(val):
                if "silver" in key_l:
                    return int(val)
            result = search_json_for_silver_price(val, depth + 1)
            if result:
                return result
    elif isinstance(obj, list):
        for item in obj:
            result = search_json_for_silver_price(item, depth + 1)
            if result:
                return result
    return None


def parse_anchored_text(html):
    """Legacy strategy: anchor on 'SILVER ... per 1 tola', with a fallback
    that counts to the third 'per 1 tola' occurrence (Fine Gold, Tejabi,
    Silver, in that order on the page)."""
    text = strip_html(html)

    m = re.search(
        r"SILVER\s*per\s*1\s*tola[\s\S]{0,200}?(?:[\u0930\u0941]+|Nrs)\s*\*{0,2}(\d{3,6})(?:\.\d+)?\*{0,2}",
        text, flags=re.I,
    )
    if m:
        price = int(m.group(1))
        if in_range(price):
            return price

    tola_matches = list(re.finditer(r"per\s*1\s*tola", text, flags=re.I))
    if len(tola_matches) >= 3:
        idx = tola_matches[2].end()
        window = text[idx: idx + 200]
        wm = re.search(r"(?:[\u0930\u0941]+|Nrs)\s*\*{0,2}(\d{3,6})(?:\.\d+)?\*{0,2}", window, flags=re.I)
        if wm:
            price = int(wm.group(1))
            if in_range(price):
                return price

    return None


def source_fenegosida():
    for name, url in build_proxy_targets():
        print(f"🌐 [fenegosida.org via {name}] fetching")
        html = fetch_url(url, retries=1 if name != "Direct" else 2)
        if not html:
            continue
        print(f"   📄 page size: {round(len(html) / 1024)}KB")

        if "SILVER" not in html.upper():
            print("   ⚠️  no 'SILVER' text present — likely the client-rendered empty shell")
            continue

        price = parse_embedded_json(html)
        if price:
            print(f"   ✅ price found via embedded JSON state: {price}")
            return price

        price = parse_anchored_text(html)
        if price:
            print(f"   ✅ price found via anchored text parse: {price}")
            return price

        print("   ⚠️  page had 'SILVER' text but no parseable price")
    return None


# ───────────────────────── Source 2: kokil.com.np mirror ─────────────────────────
# Server-rendered page that republishes official FENEGOSIDA rates daily.
# Confirmed structure (2026-08): "Silver Chandi · Pure Silver Rs. 4,645
# per tola Rs. 3,983 / 10g"

def source_kokil():
    url = "https://kokil.com.np/tools/nepal-gold-silver-price"
    print(f"🌐 [kokil.com.np] fetching {url}")
    html = fetch_url(url)
    if not html:
        return None
    print(f"   📄 page size: {round(len(html) / 1024)}KB")

    text = strip_html(html)

    # Anchor on "Pure Silver" specifically — not just "Silver", which also
    # appears in the page's "Gold & Silver Rate" title far above the real
    # data block and would otherwise match the wrong (gold) price.
    candidates = list(re.finditer(
        r"Pure Silver.{0,80}?Rs\.?\s*([\d,]{3,7})\s*per\s*tola",
        text, flags=re.I | re.S,
    ))
    # Fallback anchor in case the "Pure Silver" wording ever changes.
    if not candidates:
        candidates = list(re.finditer(
            r"Silver.{0,80}?Rs\.?\s*([\d,]{3,7})\s*per\s*tola",
            text, flags=re.I | re.S,
        ))

    # Check every match in order, not just the first — an early match that
    # falls outside the plausible range shouldn't stop us from trying the
    # rest of the matches further down the page.
    for m in candidates:
        price = int(m.group(1).replace(",", ""))
        if in_range(price):
            print(f"   ✅ Silver per-tola price found: {price}")
            return price
        print(f"   ⚠️  candidate {price} out of plausible range, trying next match")

    print("   ⚠️  no price found in this response")
    return None


SOURCES = [
    ("fenegosida.org", source_fenegosida),
    ("kokil.com.np (FENEGOSIDA mirror)", source_kokil),
]


def get_silver_price():
    for name, fn in SOURCES:
        try:
            price = fn()
        except Exception as e:  # noqa: BLE001 - one bad source must not kill the run
            print(f"   ❌ source '{name}' raised an unexpected error: {e}")
            price = None
        if price:
            return price, name
    return None, None


# ───────────────────────── Firestore ─────────────────────────

def firestore_request(path, method, fields):
    body = json.dumps({"fields": fields}).encode("utf-8")
    url = f"https://firestore.googleapis.com/v1/projects/{PROJECT_ID}/databases/(default)/documents/{path}?key={API_KEY}"
    req = urllib.request.Request(url, data=body, method=method, headers={
        "Content-Type": "application/json",
        "Content-Length": str(len(body)),
    })
    with urllib.request.urlopen(req, timeout=15) as resp:
        return resp.status, resp.read().decode("utf-8")


def write_to_firestore(price, source_name):
    fields = {
        "rate": {"integerValue": str(price)},
        "updatedBy": {"stringValue": "auto@github-actions"},
        "lastUpdated": {"stringValue": datetime.now(timezone.utc).isoformat()},
        "source": {"stringValue": source_name},
        "note": {"stringValue": "Silver per tola"},
    }
    status, data = firestore_request("global_data/silver_info", "PATCH", fields)
    if status != 200:
        raise RuntimeError(f"Firestore error {status}: {data}")
    return data


def log_to_history(price, source_name):
    fields = {
        "rate": {"integerValue": str(price)},
        "updatedBy": {"stringValue": "auto@github-actions"},
        "timestamp": {"stringValue": datetime.now(timezone.utc).isoformat()},
        "source": {"stringValue": source_name},
        "note": {"stringValue": "Silver per tola"},
    }
    try:
        status, data = firestore_request("silver_history", "POST", fields)
        if status not in (200, 201):
            print(f"⚠️  history log returned status {status}: {data}")
        else:
            print("📋 Silver logged to silver_history collection")
    except Exception as e:  # noqa: BLE001 - history logging is best-effort
        print(f"⚠️  history log skipped: {e}")


# ───────────────────────── Main ─────────────────────────

def main():
    print("🥈 Silver price auto-fetch starting...")
    print(f"📅 Time (UTC): {datetime.now(timezone.utc).isoformat()}")
    print(f"🌐 Source chain: {' → '.join(name for name, _ in SOURCES)}")

    price, source_name = get_silver_price()
    if not price:
        print("❌ All sources exhausted. Could not obtain a silver price.")
        sys.exit(1)

    print(f"🥈 Silver price (per tola): NPR {price:,}  [source: {source_name}]")

    try:
        write_to_firestore(price, source_name)
        print("🔥 Silver written to Firestore → global_data/silver_info")
    except Exception as e:
        print(f"❌ Firestore write failed: {e}")
        sys.exit(1)

    log_to_history(price, source_name)

    print()
    print("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
    print(f"✅ Done! silver_info.rate = {price}")
    print("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")


if __name__ == "__main__":
    main()
